Added a command to /eco that allows users to take percentages from pl…

CrashCringle12 opened the pull request:
…ayers

Command syntax: /eco takeper <player> <amount>

Updated messages and command desc accordingly.

Ichbinjoe reviewed (changes requested):
Why a new command? Why not just 'overload' take?

CrashCringle12 commented:
I think if take was hypothetically overloaded, give should also receive the same.

I figured taking percentages from someone's balance could quickly get real complicated for that edge case where you have to use the command multiple times. Accidentally forgetting the percent sign might cause problems and consequently as would not noticing. My thought process was have the second command so there's a clear distinction between the two and if the executor types the percent command it's more likely that it was intentional.

Hmm thoughts?

pop4959 commented:
In my opinion "takeper" just sounds silly (just go with takepercent if you must).

More seriously though, I don't think it's that easy to accidentally type the percent symbol. If you're typing in numbers, chances are you are not holding shift. Also, if you are adding the ability to take a percent, it would also make sense to include the functionality for give and set as well. As is, implementing it the way you are is duplicating a lot of code.

Ichbinjoe commented:
Agreed - if one is merged so should the other.

I agree with @pop4959 - you are basically duplicating all functionality besides the number / value parsing. I don't believe that the additional % would result in mistakes - I would expect people are fairly diligent about being careful when shuffling virtual currency around. UI > safety in this instance.

CrashCringle12 commented:
Hmm alright those are actually some good points. I'll make the change

CrashCringle12 commented:
Hm so I removed the new command and condensed things as much as possible. I added one method to handle percentages with take/give. 

I'm unsure why all of my builds are failing tho, I haven't modified anything aside from these same files, any idea what's happening here?

pop4959 commented:
It's not your fault, Spigot added some new methods that is causing Essentials tests to fail:
```
[ERROR] /home/runner/work/Essentials/Essentials/Essentials/test/com/earth2me/essentials/FakeServer.java:[517,47] <anonymous com.earth2me.essentials.FakeServer$2> is not abstract and does not override abstract method setStatistic(org.bukkit.Statistic,org.bukkit.entity.EntityType,int) in org.bukkit.OfflinePlayer
```

I'll force a rebuild on this PR later so you can see the real status.

CrashCringle12 commented:
Ahhh I see, ty ty.

pop4959 commented:
The build should pass once #3140 is merged.

JRoy commented:
(And once this PR rebased)

pop4959 commented:
Actually, pretty sure the diff from this PR should still work. He might need to rebase if he isn't able to build on his side, but chances are he's using a cached version of the Spigot API. If he runs maven clean and/or invalidates cache in IntelliJ he may have to do that though.

CrashCringle12 commented:
Alright this should be good, let me know what you guys think. It works for set, give, and take with its own message for each.
And i didn't feel like waiting for the pull request (lol) so i just updated FakeServer to get the builds working.

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/User.java:
```diff
@@ -220,7 +220,34 @@ public void takeMoney(final BigDecimal value, final CommandSource initiator, Use
+        }
+        BigDecimal value = getMoney().multiply(percent.divide(new BigDecimal("100.0")));
+        try {
+            switch (cmd) {
```
Noooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooope

Begone with your magic numbers!!!!

Make this an enum or refactor out - prefer the second to the first but :man_shrugging:

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -35,7 +36,13 @@ public void run(final Server server, final CommandSource sender, final String co
         try {
             cmd = Commandeco.EcoCommands.valueOf(args[0].toUpperCase(Locale.ENGLISH));
-            amount = (cmd == Commandeco.EcoCommands.RESET) ? startingBalance : new BigDecimal(args[2].replaceAll("[^0-9\\.]", ""));
+            if (args[2].contains("%") && !(cmd == Commandeco.EcoCommands.RESET)) {
+                isPercent = true;
+                amount = new BigDecimal(args[2].substring(0,args[2].length()-1)); 
```
This is not sane - this will pass 5%3, and will then throw an exception because it can't parse the number.... which is then sent back to the user as 'NotEnoughArgumentsException'.

CrashCringle12 replied:
Oof alright alright, fair. That is pretty bonkers now that you mention it. I'll check if the last character is '%' rather than if the entire string contains one.

```suggestion
if (args[2].endsWith("%") && (cmd != Commandeco.EcoCommands.RESET))
```

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -55,13 +62,15 @@ public void run(final Server server, final CommandSource sender, final String co
             case GIVE:
-                player.giveMoney(amount, sender, UserBalanceUpdateEvent.Cause.COMMAND_ECO);
+                if (isPercent)
+                    player.percentHandler(amount, sender, UserBalanceUpdateEvent.Cause.COMMAND_ECO, 0);
```
Please refactor into something along the lines of 'giveMoneyPercent' instead of using these magic values and a switch.

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -91,10 +104,14 @@ private void set(BigDecimal amount, final User player, final CommandSource sende
-        if (sender != null) {
-            sender.sendMessage(tl("setBalOthers", player.getDisplayName(), NumberUtil.displayCurrency(player.getMoney(), ess)));
+        if (isPercent) {
+            player.percentHandler(amount, sender, UserBalanceUpdateEvent.Cause.COMMAND_ECO, 2);
```
This implementation doesn't line up with the implementation below... I get its non-congruent with the above two methods but this way makes it look like the messages are missing.

Ichbinjoe reviewed (changes requested):
I think I see what you were trying to achieve - I think it would be simpler if there were simple forks in the code which that when a percentage was encountered would reset amount to the proportional amount (depending on context) and continue on the original execution - I don't think any value is lost by not telling them the particular percentage.

CrashCringle12 commented:
Hm alright, so where it stands now percentages are handled properly and everything extra except for the messages I added have been yeeted.

It feels like percentage should be displayed though. Whenever I use the command there's a lack of confirmation I guess? I mean it tells me the balance taken away sure, but I think seeing the percent would be best.

I have a number of different ideas for how to make it work, but I'd be taking up a similar method to that of my original commit (minus the extra command ofc). Am I better off making additional methods in User or within Eco? As I'd be making 3 mostly copied new methods or 1 method that uses an ENUM to tell whether it's take, give, or set. Or i mean i could modify take and add a give, and set method to Eco. Just sort of brain dumping the ideas i have rn before I make another attempt at it. Unless you guys think it's really fine without the message

Ichbinjoe commented:
@pop4959 for opinion about message.

New code should follow the convention of the code which already exists. Enum to define behavior is usually a code smell (especially since we don't have a good reason to really use it).

pop4959 commented:
As long as there is already a confirmation for the amount removed/added to the account, I don't think it's necessary to add new messages specifically for this. You could just calculate the final amount based on the percent and let the existing code handle it. It shouldn't be too confusing for users since they can see what the current balance is, as well as what was removed from it.

CrashCringle12 commented:
Hmm I suppose, though I think it's worth mentioning tho you don't know the previous balance beforehand unless you check.
![image](https://user-images.githubusercontent.com/30600688/78639777-8d4af880-787c-11ea-82a3-f08c36353fb3.png)
I definitely see where you're saying tho and understand, I mean it makes sense... just I always operate on the notion that people can always be numskulls lmao.

I removed the messages and just left the percent functionality. I'm alright compromising on that. How am I looking now?

pop4959 commented:
Well, people can be stupid any way you cut it. Since the % thing is kind of an easter egg here, it shouldn't be much of a problem anyways. It's also hard to accidentally type, as aforementioned.

Also you actually do know the prior balance.

New balance = Starting balance - Balance removed (e.g. $900 = $1000 - $100 (10%))
->
Starting balance = New balance + Balance removed (e.g. $1000 = $900 + $100)

Simple arithmetic that can be worked out, if you make a mistake. Knowing the percent you typed in isn't even necessary, but you could use chat history to check that as well.

pop4959 commented on Essentials/src/com/earth2me/essentials/User.java:
```diff
@@ -220,7 +220,7 @@ public void takeMoney(final BigDecimal value, final CommandSource initiator, Use
         }
     }
-
+    
```
Try to avoid unnecessary formatting changes, as this results in a dirty diff.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -35,7 +36,9 @@ public void run(final Server server, final CommandSource sender, final String co
             cmd = Commandeco.EcoCommands.valueOf(args[0].toUpperCase(Locale.ENGLISH));
+            isPercent = (args[2].endsWith("%"));
             amount = (cmd == Commandeco.EcoCommands.RESET) ? startingBalance : new BigDecimal(args[2].replaceAll("[^0-9\\.]", ""));
+            
```
=

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -61,7 +67,7 @@ protected void updatePlayer(final Server server, final CommandSource sender, fin
                 take(amount, player, sender);
                 break;
-
+                
```
=

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -73,7 +79,7 @@ private void take(BigDecimal amount, final User player, final CommandSource send
         BigDecimal minBalance = ess.getSettings().getMinMoney();
         if (money.subtract(amount).compareTo(minBalance) >= 0) {
-            player.takeMoney(amount, sender, UserBalanceUpdateEvent.Cause.COMMAND_ECO);
+                player.takeMoney(amount, sender, UserBalanceUpdateEvent.Cause.COMMAND_ECO);
```
=

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -91,12 +97,12 @@ private void set(BigDecimal amount, final User player, final CommandSource sende
-        player.sendMessage(tl("setBal", NumberUtil.displayCurrency(player.getMoney(), ess)));
-        if (sender != null) {
-            sender.sendMessage(tl("setBalOthers", player.getDisplayName(), NumberUtil.displayCurrency(player.getMoney(), ess)));
+            player.setMoney(underMinimum ? minBalance : aboveMax ? maxBalance : amount, UserBalanceUpdateEvent.Cause.COMMAND_ECO);
```
= (all of this under this line)

pop4959 commented on Essentials/src/messages.properties:
```diff
@@ -379,7 +379,7 @@ onlyPlayerSkulls=\u00a74You can only set the owner of player skulls (\u00a7c397\
 openingDisposal=\u00a76Opening disposal menu...
 orderBalances=\u00a76Ordering balances of\u00a7c {0} \u00a76users, please wait...
-oversizedMute=�4You may not mute a player for this period of time.
+oversizedMute=\u00a74You may not mute a player for this period of time.
```
This isn't relevant to this PR. If there is an issue with this message, please open another PR.

pop4959 commented:
Just clean up this PR and show that you've tested the commands and we're probably good.

CrashCringle12 committed:
Removed unnecessary formatting

CrashCringle12 commented:
![image](https://user-images.githubusercontent.com/30600688/79024807-0afb5680-7b52-11ea-86e8-95b2d9657732.png)
![image](https://user-images.githubusercontent.com/30600688/79024874-3a11c800-7b52-11ea-85e8-624ba1a7db16.png)
![image](https://user-images.githubusercontent.com/30600688/79024900-4d249800-7b52-11ea-8da9-c4e86566a46d.png)

CrashCringle12 commented:
Alright how we looking

CrashCringle12 committed:
Merge branch '2.x' of https://github.com/EssentialsX/Essentials into 2.x

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -20,7 +20,7 @@
     Commandeco.EcoCommands cmd;
     BigDecimal amount;
-
+	boolean isPercent;
```
This should be 4 spaces to remain consistent with the code formatting in this file.

CrashCringle12 replied:
Gotcha

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -35,6 +35,7 @@ public void run(final Server server, final CommandSource sender, final String co
 
         try {
             cmd = Commandeco.EcoCommands.valueOf(args[0].toUpperCase(Locale.ENGLISH));
+            isPercent = (args[2].endsWith("%"));
```
Don't need parentheses here.

pop4959 commented:
@CrashCringle12 Sorry for the long wait, forgot to come back to this issue. Minor changes to be made, but the diff is way wayyy cleaner than last time, overall. I'm not really a big fan of global variables floating around personally, but since the command seems to use them anyway for the command type and amount, I don't see the harm.

@Ichbinjoe How're you feeling about this?

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -53,6 +54,9 @@ public void run(final Server server, final CommandSource sender, final String co
 
     @Override
     protected void updatePlayer(final Server server, final CommandSource sender, final User player, final String[] args) throws NotEnoughArgumentsException, ChargeException, MaxMoneyException {
+        if(isPercent && cmd != EcoCommands.RESET) {
```
Forgot, there should also be a space between the `if` and `(` here. By the way, using CTRL+ALT+L in IntelliJ will auto format the file for you, in standard formatting (which is what is used here).

Ichbinjoe commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -53,6 +54,9 @@ public void run(final Server server, final CommandSource sender, final String co
     @Override
     protected void updatePlayer(final Server server, final CommandSource sender, final User player, final String[] args) throws NotEnoughArgumentsException, ChargeException, MaxMoneyException {
+        if(isPercent && cmd != EcoCommands.RESET) {
+            amount = player.getMoney().multiply(amount.divide(new BigDecimal("100.0")));
```
This can be optimized quite a bit.
```suggestion
            amount = player.getMoney().multiply(amount).scaleByPowerOfTen(-2);
```

This abuses how BigDecimals & Integers work to get around a division.

CrashCringle12 committed:
Merge branch '2.x' of https://github.com/EssentialsX/Essentials into 2.x

CrashCringle12 committed:
Fixed spacing around isPercent, removed unnecessary parenthesis, optimized conversion

CrashCringle12 commented:
You're all good, take your time. Took me a while to respond myself.
I fixed the changes you guys suggested !
